pkg/database/all_ops.go

```diff
@@ -82,29 +88,39 @@ func (d *db) ExecAll(req *schema.ExecAllRequest) (*schema.TxHeader, error) {
 					return nil, store.ErrIllegalArguments
 				}
 
-				// check key does not exists or it's already a reference
-				entry, err := d.getAt(EncodeKey(x.Ref.Key), 0, 0, index, tx)
-				if err != nil && err != store.ErrKeyNotFound {
-					return nil, err
-				}
-				if entry != nil && entry.ReferencedBy == nil {
-					return nil, ErrFinalKeyCannotBeConvertedIntoReference
+				if req.NoWait && (x.Ref.AtTx != 0 || !x.Ref.BoundRef) {
+					return nil, ErrNoWaitOperationMustBeSelfContained
```

Comment:
Maybe we should add some more descriptive info here:

```suggestion
					return nil, fmt.Errorf(
					    "%w: can only set references to keys added within same transaction, please use bound references with AtTx set to 0",
					    ErrNoWaitOperationMustBeSelfContained,
					)
```